### PR TITLE
[Feature] Add search request history and dialog to manager dashboard

### LINKED ISSUE #11346: ✨ Search request history on Manager dashboard
## ✨ Feature

The manager dashboard needs to be able to load a list of the manager's search requests and open a preview dialog.

## 🎨 Design

@JoshBeveridge 

[Manager dashboard page](https://www.figma.com/design/oZhQe9ahidHFfaiqgREvxa/Dashboard-(All-users)?node-id=1010-18678&t=vrILuDBzHFB3b4Gq-4)

[Dialog view with mobile + dark/light mode views](https://www.figma.com/design/hPPZeQWidbmcapHl0IKhD8/Talent-requests-(Managers)?node-id=1-2&node-type=CANVAS&t=BgoJf1374u5gLEEf-0)

## 📸 Screenshot

image

image

## 🙋‍♀️ Proposed Implementation

This will require adding a method of getting "my requests" from the API. Maybe a "talentRequests" or "talentRequestHistory" field on the User object.

```graphql
// Schema.graphql
type User {
.
.
.
talentRequest: PoolCandidateSearchRequest!
}
```
```php
// PoolCandidateSearchResult.php
public function getCandidateCountAttribute()
{
    $applicantFilter = $this->applicantFilter();
    $candidateCount = CountPoolCandidatesByPool::candidateCount($applicantFilter);
    return $candidateCount;
}
```

For the sorting, that can be done on the front-end similar to ExperienceSortAndFilter.tsx

### Status badge table
| Enum               | EN                | FR |
|--------------------|-------------------|----|
| NEW                | Submitted         |    |
| IN_PROGRESS        | Under review      |    |
| WAITING            | Awaiting response |    |
| DONE               | Complete          |    |
| DONE_NO_CANDIDATES | Complete          |    |
| NOT_COMPLIANT      | Complete                  |    |

## ✅ Acceptance Criteria

- [ ] Schema is updated to allow a manager to requests their search requests with status, potential match count, and opening date
- [ ] API offers a way for a candidate to load their history of talent request. Since this request can be expensive on the server, let's separate it into two. First request for the talent requests without the candidate count. Second, another request for the candidate count.
- [ ] Create a sorting feature using similar method as [ExperienceSortAndFilter.tsx ](https://github.com/GCTC-NTGC/gc-digital-talent/blob/main/apps/web/src/components/ExperienceSortAndFilter/ExperienceSortAndFilter.tsx#L1-L136)
- [ ] permissions ensure a user (even an admin) cannot view another user's request history
- [ ] On page load, a list of the user's search requests is queried from the API and displayed
- [ ] clicking the preview button opens the preview dialog
- [ ] TODO: clicking the "view request" button in the preview does what? A
       ANSWER: Omit the "View request" button in the footer. There should only be the "Close" button.

## 🛑 Blockers

```[tasklist]
### Blocked By
- [ ] https://github.com/GCTC-NTGC/gc-digital-talent/issues/10785
```

### PR DESCRIPTION
**Pull request by petertgiles**
🤖 Resolves #11346 

## 👋 Introduction

This branch adds the ability to load requests from the database and preview them in a dialog.

## 🧪 Testing

1. Comment out the loader for ../pages/ManagerDashboardPage/ManagerDashboardPage in apps/web/src/components/Router.tsx
2. .Rebuild the app
3. Log in as manager@test.com
4. Visit /en/search and submit some talent requests with various options selected
5. Visit /en/manager/dashboard and ensure the requests appear as preview items
6. Click on the preview items to ensure that the options are accurately shown

## 📸 Screenshot

image

## :truck: Deployment

Rerun the role-permission seeder:

`./artisan db:seed --class=RolePermissionSeeder`

### CODE REVIEW

File: api/app/Models/PoolCandidateSearchRequest.php [hunk 1]
```diff
@@ -67,7 +68,8 @@ protected static function boot()
 
         parent::boot();
         static::creating(function ($searchRequest) use ($user) {
-            $searchRequest->user_id = $user?->id;
+            // unless the user_id was already specified use the currently authenticated user
+            $searchRequest->user_id = $searchRequest->user_id ?? $user?->id;
```
Comment (petertgiles) [inline api/app/Models/PoolCandidateSearchRequest.php:72]: This change was needed to be able to set the user_id in a factory.

File: api/graphql/schema.graphql [hunk 2]
```diff
@@ -143,6 +143,8 @@ type User {
 
   # Bookmarks
   poolBookmarks: [Pool] @belongsToMany
+  # My talent requests
+  poolCandidateSearchRequests: [PoolCandidateSearchRequest!]
```
Comment (petertgiles) [inline api/graphql/schema.graphql:147]: No guards on this.  Assuming if you can see the user you can see their requests as well.

File: apps/web/src/components/SearchRequestFilters/SearchRequestFilters.tsx
Comment (petertgiles) [inline apps/web/src/components/SearchRequestFilters/SearchRequestFilters.tsx:1]: Just factoring stuff out so it can be reused.

**Review by petertgiles — COMMENTED** [review]
Some notes

**Comment by petertgiles** [comment]
@tristan-orourke Could I ask you to please take a look at the schema, role, and policy changes in this branch?

File: api/tests/Feature/PoolCandidateSearchRequestTest.php [hunk 3]
```diff
@@ -344,4 +344,80 @@ public function testUserIdForLoggedInUsers()
...
+
+    public function testUserCanSeeTheirOwnRequests()
+    {
+        $this->seed(DepartmentSeeder::class);
+        $this->seed(CommunitySeeder::class);
+
+        $user1 = User::factory()->create();
+        $request1 = PoolCandidateSearchRequest::factory()->create(['user_id' => $user1->id]);
+        $user2 = User::factory()->create();
+        $request2 = PoolCandidateSearchRequest::factory()->create(['user_id' => $user2->id]);
```
Comment (yonikid15) [inline api/tests/Feature/PoolCandidateSearchRequestTest.php:356]: Might be overkill, but can we create a third request with no user attached (user_id = "null") to ensure they dont show up as well?

File: apps/web/src/components/SearchRequestFilters/SearchRequestFilters.tsx [hunk 4]
```diff
@@ -172,7 +130,7 @@ const ApplicantFilters = ({
       <div data-h2-flex-item="base(1of1) p-tablet(1of2)">
         <div>
           <FilterBlock
-            title={intl.formatMessage(messages.community)}
+            title={intl.formatMessage(adminMessages.community)}
```
Comment (yonikid15) [inline apps/web/src/components/SearchRequestFilters/SearchRequestFilters.tsx:133]: Should we add a new community message to talentRequestsMessages as well? Then the component will only rely on one file. If you decide to do this then we should add stream (from processMessages.stream) as well.

File: apps/web/src/pages/ManagerDashboardPage/Components/PoolCandidateSearchRequestPreviewListItem.tsx [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+      ? formatClassificationString(classifications[0])
+      : null;
+
+  const secondPart =
+    request.jobTitle ?? intl.formatMessage(commonMessages.notProvided);
+
+  const completedTitle =
+    typeof firstPart === "string"
+      ? `${firstPart}${intl.formatMessage(commonMessages.dividingColon)}${secondPart}`
+      : secondPart;
```
Comment (yonikid15) [inline apps/web/src/pages/ManagerDashboardPage/Components/PoolCandidateSearchRequestPreviewListItem.tsx:55]: I'm not a big fan of `firstPart` and `secondPart`. I think just calling it what the variable would be a better option: `classification` and `requestJobTitle`.

File: apps/web/src/pages/ManagerDashboardPage/Components/PoolCandidateSearchRequestPreviewListItem.tsx [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+  const title = buildTitle(request, intl);
+
+  type MetaDataProps = React.ComponentProps<
+    typeof PreviewList.Item
+  >["metaData"];
+  type MetaDataPropItem = MetaDataProps[number];
+  const metaDataProps: MetaDataPropItem[] = [];
+  if (request.status?.value) {
+    const chipSettings = deriveChipSettings(request.status?.value, intl);
```
Comment (yonikid15) [inline apps/web/src/pages/ManagerDashboardPage/Components/PoolCandidateSearchRequestPreviewListItem.tsx:86]: You can remove the null check "?" here.

```suggestion
    const chipSettings = deriveChipSettings(request.status.value, intl);
```

File: apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+          <Dialog.Footer>
+            {showUnfinishedPieces && (
+              // unfinished - need a url to a pge that doesn't exist yet, but we do have the ID already
+              <Link color="secondary" mode="solid" href="#">
+                {intl.formatMessage({
+                  defaultMessage: "View request",
+                  id: "yycjgL",
+                  description: "Button text to go to view the request",
+                })}
+              </Link>
```
Comment (yonikid15) [inline apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx:445]: I thought we decided to not do a view single talent request page right now? Instead were going to remove the view request button. Sorry if the issue didn't mention this. That's on me.

File: apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+            }
+          </FieldDisplay>
+        </div>
+        <Separator orientation="horizontal" data-h2-margin="base(0)" />
+        <div
+          data-h2-display="base(flex)"
+          data-h2-flex-direction="base(column)"
+          data-h2-gap="base(x0.5)" // accordions have x0.5 margin built-in already
+        >
+          <Accordion.Root type="single" collapsible data-h2-margin="base(0)">
```
Comment (yonikid15) [inline apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx:258]: You can move all the accordion items under one Root and make a group. (Currently each accordion in the dialog has it's own Root)

File: apps/web/src/pages/ManagerDashboardPage/utils.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  const localizedStrings = values.map(localizedMapper);
+  localizedStrings.sort((a, b) =>
+    a.localeCompare(b, locale, { sensitivity: "base" }),
+  );
+
+  const uniqueStrings = uniq(localizedStrings);
+  const joinedStrings = uniqueStrings.join(", ");
+
+  return joinedStrings;
+}
```
Comment (yonikid15) [inline apps/web/src/pages/ManagerDashboardPage/utils.ts:59]: This looks very useful! Maybe we should add it to packages/helpers/utils/util?

**Review by yonikid15 — CHANGES_REQUESTED** [review]
Looking really good Peter! Nice work!
Just found  a few things that could need tweaking.

File: apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx (hunk 8 above)
Comment (petertgiles) [reply apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx:258]: It makes sense to me but I'm not sure that's how the design has them.  You can see each collapsible is at the same nesting level as the neighboring elements and the spacing is noticeably larger than our accordion group. 
image

@JoshBeveridge can you confirm?

File: apps/web/src/pages/ManagerDashboardPage/utils.ts (hunk 9 above)
Comment (petertgiles) [reply apps/web/src/pages/ManagerDashboardPage/utils.ts:59]: I'm glad you like it!  :smile: In the later PR I pull it out a level to share between manager pages.  I'm not sure I want to preemptively pull it out further but if you end up using it at a wider level, by all means factor it out to the helper package at that point.

File: apps/web/src/components/SearchRequestFilters/SearchRequestFilters.tsx (hunk 4 above)
Comment (petertgiles) [reply apps/web/src/components/SearchRequestFilters/SearchRequestFilters.tsx]: It's a bit tidier, nice!
[copy messages](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11653/commits/78728319dd85a7a3cef18f87095b97c6fe6b70e9)

File: apps/web/src/pages/ManagerDashboardPage/Components/PoolCandidateSearchRequestPreviewListItem.tsx (hunk 6 above)
Comment (petertgiles) [reply apps/web/src/pages/ManagerDashboardPage/Components/PoolCandidateSearchRequestPreviewListItem.tsx]: good catch!
[remove unnecessary nullsafe dereference](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11653/commits/cbe68f2195c5807fb2f032fa5c6dd3664e21fb59)

File: apps/web/src/pages/ManagerDashboardPage/Components/PoolCandidateSearchRequestPreviewListItem.tsx (hunk 5 above)
Comment (petertgiles) [reply apps/web/src/pages/ManagerDashboardPage/Components/PoolCandidateSearchRequestPreviewListItem.tsx]: There's some room for improvement.  :sweat_smile: 
[improve variable names](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11653/commits/b15f6c24593293e81479afe992d072a972de1e68)

File: apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx (hunk 7 above)
Comment (petertgiles) [reply apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx]: OK, if we're settled on that.  It means I don't have to drill that "unfinished pieces" prop through, too.
[remove view request button](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11653/commits/ab5f22a03c54fb47aba7ac056eb59c1e9881be59)

File: apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx (hunk 8 above)
Comment (JoshBeveridge) [reply apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx:257]: @petertgiles I hadn't used a group here, but there are standards outlined in the accordion component Figma page for how groups should be structured and spaced. Best to follow that. Let me know if it diverges drastically from what you're seeing in this dialog and I'll take another look.

Comment (petertgiles) [reply apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx:257]: OK, sounds good, I'll switch it over.  The accordion group is just spaced a bit tighter.
image

Comment (petertgiles) [reply apps/web/src/pages/ManagerDashboardPage/Components/ReviewTalentRequestDialog.tsx]: [switch to accordion group](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11653/commits/285d31e2c10138acd143a343f5851ab18715b621)

**Review by yonikid15 — APPROVED** [review]
Approved :heavy_check_mark:

**Comment by yonikid15** [comment]
> @tristan-orourke Could I ask you to please take a look at the schema, role, and policy changes in this branch?

@tristan-orourke could you take a look at these when you get a chance :smile:

**Comment by tristan-orourke** [comment]
^ Thanks, I somehow missed this notification last week

File: api/app/Models/PoolCandidateSearchRequest.php (hunk 1 above)
Comment (tristan-orourke) [reply api/app/Models/PoolCandidateSearchRequest.php:72]: Could you mention that in the comment, actually?

File: api/graphql/schema.graphql (hunk 2 above)
Comment (tristan-orourke) [reply api/graphql/schema.graphql:147]: I don't think this is a safe assumption, actually. We allow managers to sometimes view a candidate's "user profile". Anything extraneous to their profile data (like requests they've submitted) should be protected separately.

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Please add a check-policy directive to the schema where requests are retrieved in the User model. Besides that, it looks good!

File: api/graphql/schema.graphql (hunk 2 above)
Comment (petertgiles) [reply api/graphql/schema.graphql:146]: Good idea.  I've add a canResolved on the view policy.
[add guard to user property](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11653/commits/a6d3e32b0ac5153869fa9d20fbebc005bfdf0136)

**Comment by petertgiles** [comment]
@tristan-orourke Could you please review my update for your requested change to the schema?

**Review by tristan-orourke — APPROVED** [review]

**Pull request merged**